kotlin-runtime/ftl-plugin/src/main/kotlin/xyz/block/ftl/gradle/SchemaGenerator.kt

```diff
@@ -0,0 +1,162 @@
+          }
+
+        require(sendSchemaChannel.trySend(PullSchemaRequest()).isSuccess)
+        try {
+          for (schema in receiveSchemaChannel) {
+            generateModule(schema)
+            if (!schema.more) {
+              receiveSchemaChannel.cancel()
+              return@launch
+            }
+          }
+        } catch (e: Exception) {
+          receiveSchemaChannel.cancel()
+          throw e
+        }
+      }
+    }
+  }
+
+  private fun generateModule(schema: PullSchemaResponse) {
+    val file = FileSpec.builder("ftl.${schema.module_name}", schema.module_name.capitalized())
+      .addFileComment("Generated by FTL-Plugin")
+
+    val klass = TypeSpec.classBuilder(schema.module_name.capitalized())
+      .primaryConstructor(
+        FunSpec.constructorBuilder()
+          .build()
+      )
+
+    val verbs = schema.schema?.decls?.mapNotNull { it.verb } ?: listOf()
+    val types = schema.schema?.decls?.mapNotNull { it.data_ } ?: listOf()
+
+    types.forEach { type ->
+      val dataClassBuilder = TypeSpec.classBuilder(type.name)
+        .addModifiers(KModifier.DATA)
+        .addKdoc(type.comments.joinToString("\n"))
+
+      val constructorBuilder = FunSpec.constructorBuilder()
+      type.fields.forEach { field ->
+        dataClassBuilder.addKdoc(field.comments.joinToString("\n"))
+        field.type?.let {
+          val typeName = TypeVariableName(getTypeName(it).replace("{}", ""))
+          constructorBuilder.addParameter(field.name, typeName)
+          dataClassBuilder.addProperty(
+              PropertySpec.builder(field.name, typeName)
+                .initializer(field.name)
+                .build()
+            )
+        }
+      }
+
+      dataClassBuilder.primaryConstructor(
+        constructorBuilder
+          .build()
+      )
+
+      file.addType(
+        dataClassBuilder.build()
+      )
+    }
+
+    verbs.forEach { verb ->
+      val funSpecBuilder = FunSpec.builder(verb.name)
+        .addKdoc(verb.comments.joinToString("\n"))
+        .addAnnotation(
+          AnnotationSpec.builder(ClassName("xyz.block.ftl", "Verb"))
+            .build()
+        )
+        .addCode("throw NotImplementedError()")
+      verb.metadata.forEach { metadata ->
+        metadata.ingress?.let {
+            funSpecBuilder.addAnnotation(
+              AnnotationSpec.builder(ClassName("xyz.block.ftl", "Ingress"))
+                .addMember("%T", ClassName("xyz.block.ftl.Method", it.method))
+                .addMember("%S", it.path)
+                .build()
+            )
+        }
+      }
+
+      verb.request?.let {
+        funSpecBuilder.addParameter(
+          "req",
+          TypeVariableName(it.name)
+        )
+      }
+      verb.response?.let {
+        funSpecBuilder.returns(
+          TypeVariableName(it.name)
+        )
+      }
+
+      klass.addFunction(
+        funSpecBuilder.build()
+      )
+    }
+
+    file.addType(klass.build())
+    file.build().writeTo(System.out)
```

Comment:
Currently only writing to `System.out` but I should be able to send this to a directory as well. this was just faster for iterating on changes.